src/Workspaces/Core/Portable/Utilities/NullableHelpers/TypeSymbolWithNullableAnnotation.cs

```diff
@@ -0,0 +1,196 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Globalization;
+using System.Text;
+using System.Threading;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal static partial class NullableExtensions
+    {
+        [DebuggerDisplay("{WrappedSymbol,nq}, Nullability = {Nullability,nq}")]
+        private abstract class TypeSymbolWithNullableAnnotation : ITypeSymbol
+        {
+            internal ITypeSymbol WrappedSymbol { get; }
+            internal NullableAnnotation Nullability { get; }
+
+            protected TypeSymbolWithNullableAnnotation(ITypeSymbol wrappedSymbol, NullableAnnotation nullability)
+            {
+                Debug.Assert(!(wrappedSymbol is TypeSymbolWithNullableAnnotation));
+
+                WrappedSymbol = wrappedSymbol;
+                Nullability = nullability;
+            }
+
+            public bool Equals(ISymbol other)
+            {
+                if (other is TypeSymbolWithNullableAnnotation otherWrappingSymbol)
+                {
+                    return this.Nullability == otherWrappingSymbol.Nullability &&
+                           this.WrappedSymbol.Equals(otherWrappingSymbol.WrappedSymbol);
+                }
+                else if (other is ITypeSymbol)
+                {
+                    // Somebody is trying to compare a nullable-wrapped symbol with a regular compiler symbol. By rule Equals must be reflexive,
+                    // and since the compiler's Equals won't respect us as being equal, we can't return anything other than false. Flagging this with an assert
+                    // is helpful while moving features over, because this comparison might be the reason a feature isn't working right. However, for now disabling
+                    // the assert is easiest because we can't update the whole codebase at once. Enabling the assert is tracked in https://github.com/dotnet/roslyn/issues/36045.
+
+                    // Debug.Fail($"A {nameof(TypeSymbolWithNullableAnnotation)} was compared to a regular symbol. This comparison is disallowed.");
+
+                    // We are also going to cheat further: for now, we'll just throw away nullability and compare, because if a core feature (like the type inferrerr) is updated
+                    // but other features aren't, we want to keep those working.
+                    return this.WrappedSymbol.Equals(other);
+                }
+                else
+                {
+                    return false;
+                }
+            }
+
+            public override bool Equals(object obj)
+            {
+                return this.Equals(obj as ISymbol);
+            }
+
+            public override int GetHashCode()
+            {
+                return this.WrappedSymbol.GetHashCode();
```

Comment:
>return this.WrappedSymbol.GetHashCode(); [](start = 16, length = 40)

nit: This _seems_ wrong but I don't have a good argument for why yet. Just a code smell. I don't suppose we can use `System.HashCode.Combine(this.Nullability, this.WrappedSymbol)`?

Reply:
The evil is that right now we allow you to compare a wrapped symbol to an unwrapped symbol, and we ignore nullability in that case. Therefore, we need to ensure the in that case the hash codes are still the same. I'll add a comment.